File: Graph-Algorithm/CSHARP/strongly_connected_component.cs
using System;
using System.Collections.Generic;

public class SCC {

    public int nodes, edges;
    public List <int>[] graph;
    public List <int>[] revgraph;
    public Stack<int> st;
    public bool[] vis;

    public SCC() {

    }

    public void initSCC(int nodes, int edges) {
        this.nodes = nodes;
        this.edges = edges;

        st = new Stack<int>();
        vis = new bool[nodes];
        graph = new List<int>[nodes];
        revgraph = new List<int>[nodes];

        for (int i = 0; i < nodes; i++) {
            graph[i] = new List<int>();
            revgraph[i] = new List<int>();
        }
    } 

    public void clearVis() {
        for (int i = 0; i < nodes; i++) {
            vis[i] = false;
        }
    }

    public void dfs1(int u) {
        vis[u] = true;

        for (int i = 0; i < revgraph[u].Count; i++) {
            int v = revgraph[u][i];

            if (vis[v] == false) {
                dfs1(v);
            }
        }
        st.Push(u);
    }

    public void dfs2(int u) {
        vis[u] = true;

        Console.Write("{0} ", u);

        for (int i = 0; i < graph[u].Count; i++) {
            int v = graph[u][i];

            if (vis[v] == false) {
                dfs2(v);
            }
        }
    }

    public void manageSCC(string[] lines) {

        string[] line1 = lines[0].Split(' ');
        int nodes = int.Parse(line1[0]);
        int edges = int.Parse(line1[1]);

        initSCC(nodes, edges);

        for (int i = 1; i <= edges; i++) {
            string[] all_edge = lines[i].Split(' ');
            int u = int.Parse(all_edge[0]);
            int v = int.Parse(all_edge[1]);

            graph[u].Add(v);
            revgraph[v].Add(u);
        }

        clearVis();

        for (int i = 0; i < nodes; i++) {
            if (vis[i] == false) {
                dfs1(i);
            }
        }

        clearVis();

        while (st.Count > 0) {
            int u = st.Pop();

            if (vis[u] == false) {
                dfs2(u);
                Console.WriteLine();
            }
        }

    }
}

/*
Input:
-
5 5
1 0
0 2
2 1
0 3
3 4
Output:
-
4 
3 
0 2 1
*/

File: Graph-Algorithm/Strongly_Connected_Component.cpp
/*
  @topic      - Strongly Connected Component (SCC)
  @complexity - O(V+E)  -  Kosaraju’s algorithm
  @author     - Amirul islam
                   _
   _|_ o._ o._  __|_| _. _|_
  _>| ||| ||| |(_|| |(_|_>| |
                _|
*/

#include <bits/stdc++.h>
using namespace std;

const int mx = 1e3;

vector <int> graph[mx], revgraph[mx];

bool vis[mx];
stack <int> st;

void dfs1(int u) {
    vis[u] = 1;

    for (int i = 0; i < revgraph[u].size(); i++) {
        int v = revgraph[u][i];

        if (vis[v] == 0) {
            dfs1(v);
        }
    }

    st.push(u);
}

void dfs2(int u) {
    vis[u] = 1;
    cout << u << " ";

    for (int i = 0; i < graph[u].size(); i++) {
        int v = graph[u][i];

        if (vis[v] == 0) {
            dfs2(v);
        }
    }
}

int main() {
    int n, u, v;
    cin >> n;

    for (int i = 0; i < n; i++) {
        cin >> u >> v;
        graph[u].push_back(v);
        revgraph[v].push_back(u);
    }

    memset(vis, 0, sizeof(vis));
    for (int i = 0; i < n; i++) {
        if (vis[i] == 0) {
            dfs1(i);
        }
    }

    memset(vis, 0, sizeof(vis));
    while (!st.empty()) {
        int u = st.top();
        st.pop();

        if (vis[u] == 0) {
            dfs2(u);
            cout << "\n";
        }
    }
    return 0;
}

/*
Input:
-
5
1 0
0 2
2 1
0 3
3 4

Output:
-
4 
3 
0 2 1
*/
